### DISCUSSION TITLE
Mutating objects affect reproducibility

### DISCUSSION THREAD

**Discussion by benjreinhart** [marimo-team/marimo#1385] [General]
### Describe the bug

In [FAQ](https://docs.marimo.io/faq.html#what-problems-does-marimo-solve), Marimo states:

> In contrast, marimo guarantees that your code, outputs, and program state are consistent, eliminating hidden state and making your notebook reproducible.

I see that variables are tracked and reassignment is not allowed. However, since Python is not an immutable language, this guarantee doesn't seem possible because cells can mutate objects and that seems impossible to prevent. Consider the following example:

Cell A:
```python
foo = {"key": "value"}
print(foo["key"])
```

Cell B:
```python
print(foo["key"])
```

Cell C:
```python
foo["key"] = "value 2"
print(foo["key"])
```

Now, if I run them in order, you get what you'd expect the first time around. The problem is if I go back to Cell B and run that cell again. I now get a different value than I did the first time (see screen shot).

**The bug part of this is issue that if I re-run Cell A and then Cell B, I'm still seeing "value 2" printed.**

But the larger point of this issue is that, unless I'm missing something, it can't be the case that marimo notebooks truly guarantee consistency and reproducibility in all cases? Even if we tracked some types of modification, you could never catch them all. Code can call into other libraries that mutate (pandas, tensorflow, etc), code can dynamically eval strings as code, etc. I'd love to be wrong here because this would be awesome if we could reach the same level of guarantees that e.g. [Livebook](https://livebook.dev/) does, but I don't see how that's possible without using an immutable language.

<img width="764" alt="Screenshot 2024-05-15 at 2 33 06 PM" src="https://github.com/marimo-team/marimo/assets/606233/16f04931-dd83-458d-915d-415ede81703b">

### Environment

https://marimo.app (the online wasm version)

**Comment by akshayka** [comment]
> But the larger point of this issue is that, unless I'm missing something, it can't be the case that marimo notebooks truly guarantee consistency and reproducibility in all cases?

That's correct, and a good point. We try our best, without resorting to magic (ie, no runtime tracing, which is a losing battle), and are currently happy with being substantially more computationally reproducible than vanilla IPyKernel.

Our prescription to our users is: don't mutate variables across multiple cells. If you do, bets on "computational reproducibility" are off, unless you carefully manage the DAG yourself. You can find such language in our [docs](https://docs.marimo.io/guides/best_practices.html) and tutorials.

In this way we are similar to [`Pluto.jl`](https://github.com/fonsp/Pluto.jl?tab=readme-ov-file#dynamic-environment), which is our main inspiration. Pluto.jl takes reproducibility one step further than us by implementing its own package management system, which we have not yet done.

Happy to continue chatting. I will convert this to a Discussion because it's not bug, rather a design choice we're committed to.

EDIT: Rereading the FAQ -- I can make the claim more clear.

**Reply by akshayka** [reply]
By the way, I've been meaning to try Elixir/Livebook. I have heard great things. Thanks for the reminder!

**Reply by benjreinhart** [reply]
Elixir is the GOAT IMO, and livebook rules.

https://www.youtube.com/watch?v=pas9WdWIBHs (may be of interest)

**Comment by benjreinhart** [comment]
This all sounds good, and I didn't see the best practices 🙏🏻 

The only thing I'll add here is that I think there is still a bug... When I re-run the first cell (Cell A) and then Cell B (after running Cell C), shouldn't the printed value be reset back to `value`?

**Reply by akshayka** [reply]
If I run `A`, `B`, then `C`, then re-run `A` then `B`, the output of `B` that I see is `value`. Maybe I'm misunderstanding?

**Reply by benjreinhart** [reply]
Run `A`->`B`->`C`, then run `B` (now see `value 2`), then run `A` (I'd expect it to be `value` again now), then run `B` (still `value 2`)

![CleanShot 2024-05-15 at 15 31 05](https://github.com/marimo-team/marimo/assets/606233/fa837313-3c6a-4786-acb2-ad0bceaf9523)

**Reply by akshayka** [reply]
Oh I see. When `A` is run, both `B` and `C` are run, since they both reference `foo`. But `B` and `C` are not ordered with respect to each other (right now it's unspecified how ties are broken), and in your case it looks like `C` is running before `B`. I can see how this is confusing though

**Reply by benjreinhart** [reply]
> right now it's unspecified how ties are broken

Is there a reason not to run them in the order they are listed on the page? In this case at least, it would prevent the confusing behavior, though maybe there are other scenarios where it wouldn't help

**Reply by dmadisetti** [reply]
So related: https://github.com/marimo-team/marimo/issues/1142

Even beyond cloning, though, python can hook into C internals and create state that can't be easily touched- so I think there's always going to be some responsibility on the user.

**Reply by akshayka** [reply]
> Is there a reason not to run them in the order they are listed on the page? In this case at least, it would prevent the confusing behavior, though maybe there are other scenarios where it wouldn't help

We could ... although right now the kernel doesn't know the order of the cells on the page. Though that could be fixed.

We'd prefer that users don't rely on page order, since we may also add an (opt-in) parallelizing executor.

**Comment by dmadisetti** [comment]
@benjreinhart PR #1580 addresses this to a larger extent. You really have to go out of your way to mutate things- but of course, it is limited by Python itself.

**Reply by dmadisetti** [reply]
@benjreinhart #1580 is now merged, you can enable it by adding

```toml
[experimental]
execution_type = "strict"
```

To your `.marimo.toml` config!

![image](https://github.com/marimo-team/marimo/assets/2689338/6c9fd1eb-df9f-4944-a7a1-ba40c7a7ddfa)
